src/backend/distributed/executor/executor.c

```diff
@@ -530,17 +530,47 @@ CreateDistributedExecution(DistributedPlan *distributedPlan,
 
 
 /*
- * StartDistributedExecution opens connections for distributed execution and
- * assigns each task with shard placements that have previously been modified
- * in the current transaction to the connection that modified them.
+ * StartDistributedExecution takes the steps required to safely do a distributed
+ * execution:
+ *
+ * 1. Take necessary executor locks
+ * 2. Determine whether the transaction needs rollback, 2PC
+ * 3. Record whether a relation is accessed over multiple connections
+ * 4. Assign tasks to connections that modified the shards in the current transaction
  */
 void
 StartDistributedExecution(DistributedExecution *execution)
 {
 	DistributedPlan *distributedPlan = execution->plan;
+	CmdType operation = distributedPlan->operation;
 	Job *job = distributedPlan->workerJob;
 	List *taskList = job->taskList;
 
+	if (operation == CMD_INSERT || operation == CMD_UPDATE || operation == CMD_DELETE)
```

Comment:
will change this to `DistributedPlanModifiesDatabase` once it is merged.

Reply:
I think we should also add `|| operation == CMD_SELECT` because of `SELECT FOR UPDATE` on reference tables acquire exclusive locks (and in MX does some more stuff as well). I'm actually disappointed that none of the tests failed because of this.

I think that's because of using [replace_isolation_tester_func](https://github.com/citusdata/citus/pull/2197/files#diff-09efd54a1e80012e8973bb60022f196fR4) which switches sessions because of the locks on the workers, not on the coordinator. I think we need that for distributed tables since they do not acquire executor locks on `SELECT FOR UPDATE`, but, it also works for reference tables, which hides the missing executor locks on the coordinator.

A (semi)-related note: On the master branch, we're treating `select for update` on reference tables as modification in the sense that we're running the query on all the replicas. Shall we implement that in this PR? A test would be nice for that, we can use a function similar to [this](https://github.com/citusdata/citus/blob/master/src/test/regress/sql/sequential_modifications.sql#L16) to ensure that `SELECT` opens a transaction block.

Reply:
Another note:

After merging onto a temp branch, I realized that one of the tests is failing [here](https://github.com/citusdata/citus/blob/master/src/test/regress/expected/multi_partitioning.out#L1467-L1499).

That is basically `TRUNCATE` on a partition should acquire locks on the parent test. 

I think the main reason is that we're not acquiring these locks on `TRUNCATE`. 

Once we change the check to `DistributedPlanModifiesDatabase ()` should fix that.  

Maybe we should also return true for `SELECT FOR UPDATE` in that function, or have a separate check. Just thinking about yet.